File: ai_analyst/api/models/market_data.py
"""Market data OHLCV response models (PR-CHART-1).

Flat ResponseMeta & {} pattern — same as all prior endpoints.
Spec: docs/specs/PR_CHART_1_SPEC.md §6.2
"""

from __future__ import annotations

from pydantic import BaseModel

from ai_analyst.api.models.ops import ResponseMeta


class Candle(BaseModel):
    """Single OHLCV candle in lightweight-charts native format."""

    timestamp: int  # Unix epoch seconds
    open: float
    high: float
    low: float
    close: float
    volume: float


class OHLCVResponse(ResponseMeta):
    """GET /market-data/{instrument}/ohlcv response."""

    instrument: str
    timeframe: str
    candles: list[Candle]
    candle_count: int

File: ui/src/shared/api/marketData.ts
// ---------------------------------------------------------------------------
// Market Data endpoint client — shapes from PR_CHART_1_SPEC.md §6.2.
// ---------------------------------------------------------------------------

import { apiFetch, type ApiResult } from "./client";
import type { ResponseMeta } from "./ops";

// ---- Market Data types ----

export type Candle = {
  timestamp: number; // Unix epoch seconds
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
};

export type OHLCVResponse = ResponseMeta & {
  instrument: string;
  timeframe: string;
  candles: Candle[];
  candle_count: number;
};

// ---- Endpoint function ----

export type FetchOHLCVParams = {
  instrument: string;
  timeframe?: string;
  limit?: number;
};

/** Fetch OHLCV candle data for a given instrument and timeframe. */
export function fetchOHLCV(
  params: FetchOHLCVParams,
): Promise<ApiResult<OHLCVResponse>> {
  const searchParams = new URLSearchParams();

  if (params.timeframe) searchParams.set("timeframe", params.timeframe);
  if (params.limit != null) searchParams.set("limit", String(params.limit));

  const query = searchParams.toString();
  const path = `/market-data/${encodeURIComponent(params.instrument)}/ohlcv${query ? `?${query}` : ""}`;

  return apiFetch<OHLCVResponse>(path);
}
